### PR TITLE
initial apk2 signer that shells out to apksigner

### PR DESCRIPTION
**Pull request by jvehent**
WIP

### CODE REVIEW

**Comment by g-k** [comment]
refs: https://github.com/mozilla-services/autograph/issues/64

File: monitor.go [hunk 1]
```diff
@@ -51,50 +51,63 @@ func (a *autographer) handleMonitor(w http.ResponseWriter, r *http.Request) {
...
+				sigresps[i] = formats.SignatureResponse{
+					Ref:        id(),
+					Type:       s.Config().Type,
+					Mode:       s.Config().Mode,
+					SignerID:   s.Config().ID,
+					PublicKey:  s.Config().PublicKey,
+					SignedFile: signedfile,
+					X5U:        s.Config().X5U,
+					SignerOpts: s.Config().SignerOpts,
+				}
```
Comment (jvehent) [inline monitor.go]: @g-k This is a new approach to monitor signers that only implement the FileSigner interface. It adds a new `TestFileGetter` that the signer implements to return a valid test file, in this case it's a pruned version of Focus.

We still don't have a way to verify the signed file (that would mean shelling out to apksigner in `autograph-monitor`), but at least this code exercises the signer and makes sure its configuration is good.

**Comment by g-k** [comment]
can you run `make fmt-fix` and push up the changes?

File: monitor.go (hunk 1 above)
Comment (g-k) [reply monitor.go]: WFM 

Long term, do we still want to switch the monitor to fargate (or having a separate file signer verifier) and shell out to `apksigner verify`? If so, I can make an issue for that.

File: .circleci/config.yml [hunk 2]
```diff
@@ -33,7 +33,7 @@ jobs:
   unit-test:
     docker:
       - image: circleci/golang:1.13-buster
-      - image: circleci/postgres:11
```
Comment (g-k) [inline .circleci/config.yml]: What version are we running in prod? I think it'd be good to match that.

Are we using any postgres 12 features?

File: Dockerfile [hunk 3]
```diff
@@ -9,17 +9,15 @@ RUN addgroup --gid 10001 app \
...
 # import the RDS CA bundle
 # https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/CHAP_PostgreSQL.html#PostgreSQL.Concepts.General.SSL
-RUN curl -o /tmp/rds-combined-ca-bundle.pem https://s3.amazonaws.com/rds-downloads/rds-combined-ca-bundle.pem
-RUN openssl x509 -in /tmp/rds-combined-ca-bundle.pem -inform PEM -out /usr/local/share/ca-certificates/rds-combined-ca-bundle.crt
-RUN rm -f /tmp/rds-combined-ca-bundle.pem
-RUN update-ca-certificates
+RUN curl -o /tmp/rds-combined-ca-bundle.pem https://s3.amazonaws.com/rds-downloads/rds-combined-ca-bundle.pem && \
+    openssl x509 -in /tmp/rds-combined-ca-bundle.pem -inform PEM -out /usr/local/share/ca-certificates/rds-combined-ca-bundle.crt && \
+    rm -f /tmp/rds-combined-ca-bundle.pem && \
+    update-ca-certificates
```
Comment (g-k) [inline Dockerfile]: :+1:

File: docker-compose.yml [hunk 4]
```diff
@@ -1,6 +1,8 @@
 version: "3"
 volumes:
-  tmpdir:
+  apptmpdir:
+  hsmtmpdir:
+  hsmconfig:
```
Comment (g-k) [inline docker-compose.yml]: :+1:

File: signer/apk2/apk2.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"go.mozilla.org/autograph/signer"
+)
+
+const (
+	// Type of this signer is "apk2" represents a signer that
+	// shells out to apksigner to sign artifacts
+	Type = "apk2"
```
Comment (g-k) [inline signer/apk2/apk2.go]: @jvehent what do you think about adding a version field to signers in the future?

That way we could avoid adding and removing whole signers for e.g. rsapss to genericrsa, apk to apk2, and contentsignature to contentsignaturepki. Also we could use that field to pick the underlying signing binary for the gpg2 and apk2 signers e.g. for apksigner updates in the future.

File: signer/apk2/apk2.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+)
+
+// APK2Signer holds the configuration of the signer
+type APK2Signer struct {
+	signer.Configuration
+	minSdkVersion string
```
Comment (g-k) [inline signer/apk2/apk2.go]: Is this the installed apksigner SDK version or the minimum client SDK version the signature should support?

File: signer/apk2/apk2.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+	if err != nil {
+		return nil, errors.Wrap(err, "apk2: failed to get private key from configuration")
+	}
+	switch priv.(type) {
+	case *ecdsa.PrivateKey:
+		// ecdsa is only supported in sdk 18 and higher
+		s.minSdkVersion = "18"
+	default:
+		s.minSdkVersion = "9"
+	}
```
Comment (g-k) [inline signer/apk2/apk2.go]: nit: might be useful to log which version we pick

File: tools/autograph-client/client.go [hunk 8]
```diff
@@ -508,6 +515,9 @@ func verifyMAR(signedMAR []byte) bool {
 }
 
 func verifyPGP(input []byte, signature string, pubkey string) bool {
+	return true
```
Comment (g-k) [inline tools/autograph-client/client.go]: nit: we might want to log that these are no-ops

File: tools/autograph-monitor/monitor.go [hunk 9]
```diff
@@ -158,7 +167,7 @@ func Handler() (err error) {
 		case pgp.Type, gpg2.Type:
 			// we don't verify pgp signatures because that requires building a keyring
 			// using the public key which is hard to do using the current openpgp package
-			log.Printf("Skipping verification of PGP signature from signer %q", response.SignerID)
+			log.Printf("Skipping verification of PGP signature from signer %q (we can't verify those)", response.SignerID)
```
Comment (g-k) [inline tools/autograph-monitor/monitor.go]: :+1:

File: docker-compose.yml [hunk 10]
```diff
@@ -86,16 +90,18 @@ services:
       context: tools/autograph-monitor/
     environment:
       - AUTOGRAPH_URL=http://autograph-app-hsm:8001/
+      - AUTOGRAPH_KEY=19zd4w3xirb5syjgdx8atq6g91m03bdsmzjifs2oddivswlu9qs
```
Comment (g-k) [inline docker-compose.yml]: +1 to making this explicit

**Comment by g-k** [comment]
Once this is in stage we should make sure it doesn't blow up for concurrent requests like `gpg2`

File: .circleci/config.yml (hunk 2 above)
Comment (jvehent) [reply .circleci/config.yml]: I _think_ we run 11 in prod with @Micheletto could confirm. I don't foresee using 12 features any time soon, so happy to roll it back if you think parity is more important than upgrading.

File: signer/apk2/apk2.go (hunk 6 above)
Comment (jvehent) [reply signer/apk2/apk2.go]: It's the minimum client sdk. I'll add a comment.

File: signer/apk2/apk2.go (hunk 7 above)
Comment (jvehent) [reply signer/apk2/apk2.go]: signers don't log currently, so while I agree with you, that seems like a bigger problem to solve (and one worth solving)

File: tools/autograph-client/client.go (hunk 8 above)
Comment (jvehent) [reply tools/autograph-client/client.go]: 👍

File: signer/apk2/apk2.go (hunk 5 above)
Comment (jvehent) [reply signer/apk2/apk2.go]: I need to think a little bit about how that would work. Wouldn't it make the logic inside the signer a lot more complicated though? And the tests as well?

File: monitor.go (hunk 1 above)
Comment (jvehent) [reply monitor.go]: I think that would be worthwhile. And perhaps we could also integrate the work mgoodwin is doing around xpcshell in the monitor runtime.

File: .circleci/config.yml (hunk 2 above)
Comment (g-k) [reply .circleci/config.yml]: I think it's better to test against the version we have deployed. We could test against postgres 11 and 12.

File: signer/apk2/apk2.go (hunk 5 above)
Comment (g-k) [reply signer/apk2/apk2.go]: > Wouldn't it make the logic inside the signer a lot more complicated though? And the tests as well?

More paths to cover, but less code :woman_shrugging:

I can split out another issue to discuss.

File: signer/apk2/apk2.go (hunk 7 above)
Comment (g-k) [reply signer/apk2/apk2.go]: > signers don't log currently

This is news to me and the XPI signer logs.

If there are other constraints or best practices we want to enforce like https://github.com/mozilla-services/autograph/issues/247 it'd be good to document them.

File: Makefile [hunk 11]
```diff
@@ -28,47 +44,21 @@ tag: all
...
-	$(GO) vet go.mozilla.org/autograph/signer/rsapss
+	go vet $(PACKAGES)
 
 fmt-diff:
 	gofmt -d *.go database/ signer/ tools/autograph-client/ $(shell ls tools/autograph-monitor/*.go) tools/softhsm/ tools/hawk-token-maker/ tools/make-hsm-ee/ tools/makecsr/ tools/genpki/
 
 fmt-fix:
-	gofmt -w *.go database/ signer/ tools/autograph-client/ $(shell ls tools/autograph-monitor/*.go) tools/softhsm/ tools/hawk-token-maker/ tools/make-hsm-ee/ tools/makecsr/ tools/genpki/
+	go fmt $(PACKAGES)
+	gofmt -w tools/autograph-client/ $(shell ls tools/autograph-monitor/*.go) tools/softhsm/ tools/hawk-token-maker/ tools/make-hsm-ee/ tools/makecsr/ tools/genpki/
```
Comment (g-k) [inline Makefile]: should we change this one to `go fmt` too?

File: Makefile [hunk 12]
```diff
@@ -5,6 +5,22 @@ GO := go
...
+go.mozilla.org/autograph/signer/contentsignature \
+go.mozilla.org/autograph/signer/contentsignaturepki \
+go.mozilla.org/autograph/signer/mar \
+go.mozilla.org/autograph/signer/xpi \
+go.mozilla.org/autograph/signer/mar \
+go.mozilla.org/autograph/signer/pgp \
+go.mozilla.org/autograph/signer/gpg2 \
+go.mozilla.org/autograph/signer/genericrsa \
+go.mozilla.org/autograph/signer/rsapss
+
```
Comment (g-k) [inline Makefile]: :+1: overdue

File: signer/apk2/apk2.go (hunk 7 above)
Comment (jvehent) [reply signer/apk2/apk2.go]: yeah, I don't know why I said that. we obviously log in the signers. 🤒

Comment (g-k) [reply signer/apk2/apk2.go]: np

File: .circleci/config.yml [hunk 13]
```diff
@@ -33,7 +33,7 @@ jobs:
   unit-test:
     docker:
       - image: circleci/golang:1.13-buster
-      - image: circleci/postgres:12
+      - image: circleci/postgres:10.6
```
Comment (jvehent) [inline .circleci/config.yml]: @g-k this is what we use in prod

File: Makefile (hunk 11 above)
Comment (jvehent) [reply Makefile]: go fmt only works on packages, not directories, so I didn't change it here but we could

**Pull request merged**